docs/rfcs/021-raii-runtime.md

```diff
@@ -0,0 +1,209 @@
+# Improving runtime by making use of the RAII idiom
+
+## Motivation
+
+Currently, we have several problems regarding the runtime:
+
+- `TenantState` doesn't clearly show what happens with the tenant. Repository and code comments need to describe how state transitions occur and what they mean. It is critical since the code is going to have more states. The same applies to `TimelineState`.
+- Current tenant and timeline access, shutdown design fails to shut down tenants properly, timelines, background tasks, and control their lifetimes, resulting in races.
+- `tasks_mgr` runtime is redundant, and its design allows us to have even more races without noticing.
```

Comment:
As I mentioned at the end of RFC, task runtime only wraps closure runs with `TenantId`, `TimelineId`, and `TaskKind`, and it appears we don't need this selector when we have both accessors since we store `JoinHandle`s in the tenant itself and wait till completion. Added this to the beginning of the RFC.